### ISSUE TITLE
[MGPG-59] GPG Plugin: "gpg: signing failed: Inappropriate ioctl for device"

### ISSUE THREAD

**Issue by jira-importer** [apache/maven-gpg-plugin#209]
**[Paul Hammant](https://issues.apache.org/jira/secure/ViewProfile.jspa?name=paul)** opened **[MGPG-59](https://issues.apache.org/jira/browse/MGPG-59?redirect=false)** and commented

Some versions and configurations of GPG want to pop an interactive UI, which the maven-gpg-plugin cannot handle.  This appears in the log, before Maven quits without much advice:

`*gpg: WARNING: "--no-use-agent" is an obsolete option - it has no effect`
`"gpg: signing failed: Inappropriate ioctl for device"`

!screenshot-1.png!

**The remedy is for the user to do something with gpg on the command line before going into mvn-release:prepare**. Something that will allow gpg to remember (for a period of time) the passphrase entered in that interactive UI, that the maven invocation of gpg can take advantage of afterwards.

Historically there was a gpg option '--no-use-agent', but not all gpg implementations support that today, not will going forwards as it is being removed.

Suggestion: On command error, System.out.println the remedy above, perhaps even suggesting the command to run:

`gpg --use-agent --armor --detach-sign  --output $(mktemp) pom.xml`

---

**Affects:** 1.6

**Attachments:**
- [screenshot-1.png](https://issues.apache.org/jira/secure/attachment/12873409/screenshot-1.png) (_25.17 kB_)

**Issue Links:**
- [MGPG-76](https://issues.apache.org/jira/browse/MGPG-76) need different defaults for mac os
 (_**"is duplicated by"**_)
- [HBASE-18585](https://issues.apache.org/jira/browse/HBASE-18585) Update release instructions involving gpg-agent

2 votes, 9 watchers

**Comment by jira-importer** [comment]
**[Ulises Cervino Beresi](https://issues.apache.org/jira/secure/ViewProfile.jspa?name=ulises.cervino)** commented

This proposed remedy doesn't work for build servers, e.g. jenkins.

A better approach is to use `--pinentry-mode loopback` once the keys have been loaded in the agent.

**Comment by jira-importer** [comment]
**[Josef Ludvíček](https://issues.apache.org/jira/secure/ViewProfile.jspa?name=ludvicekj)** commented

Spotted this error on Mac OS 10.12.6 with

```
$ gpg --version
gpg (GnuPG) 2.2.1
libgcrypt 1.8.1

$ mvn -version
Apache Maven 3.5.2 (138edd61fd100ec658bfa2d307c43b76940a5d7d; 2017-10-18T09:58:13+02:00)
Maven home: /usr/local/Cellar/maven/3.5.2/libexec
Java version: 1.8.0_151, vendor: Oracle Corporation
Java home: /Library/Java/JavaVirtualMachines/jdk1.8.0_151.jdk/Contents/Home/jre
Default locale: en_US, platform encoding: UTF-8
OS name: "mac os x", version: "10.12.6", arch: "x86_64", family: "mac"
```

**Comment by jira-importer** [comment]
**[Robert Scholte](https://issues.apache.org/jira/secure/ViewProfile.jspa?name=rfscholte)** commented

Fixed in [11aca384c4005747d797dcc4857280d0d578d05f](https://gitbox.apache.org/repos/asf?p=maven-gpg-plugin.git;a=commit;h=11aca384c4005747d797dcc4857280d0d578d05f)
